engine/action_contact.go

```diff
@@ -4,60 +4,137 @@ import (
 	"context"
 	"errors"
 	"fmt"
+
 	"github.com/bitcoin-sv/go-paymail"
-	"github.com/mrz1836/go-cachestore"
+	"github.com/bitcoin-sv/spv-wallet/engine/utils"
 	"github.com/mrz1836/go-datastore"
 )
 
-func (c *Client) NewContact(ctx context.Context, fullName, paymail, senderPubKey string, opts ...ModelOps) (*Contact, error) {
-	// Check for existing NewRelic transaction
-	ctx = c.GetOrStartTxn(ctx, "new_contact")
+var (
+	ErrInvalidRequesterXpub = errors.New("invalid requester xpub")
+	ErrAddingContactRequest = errors.New("adding contact request failed")
+)
 
-	contact, err := getContact(ctx, fullName, paymail, senderPubKey, c.DefaultModelOptions()...)
+func (c *Client) UpsertContact(ctx context.Context, ctcFName, ctcPaymail, requesterXpub string, opts ...ModelOps) (*Contact, error) {
+	reqXPubID := utils.Hash(requesterXpub)
 
-	if contact != nil {
-		return nil, errors.New("contact already exists")
-	}
+	reqPms, err := c.GetPaymailAddressesByXPubID(ctx, reqXPubID, nil, nil, nil)
 	if err != nil {
 		return nil, err
 	}
+	if len(reqPms) == 0 {
```

Comment:
I don't think that anyone would like to request PMS:
PMS – premenstrual syndrome

![troll](https://www.icegif.com/wp-content/uploads/icegif-2456.gif)

Reply:
changed